// @amcharts/amcharts3-angular/esm2015/lib/amcharts3-angular.module.js
import { NgModule } from '@angular/core';
import { AmChartsDirective } from './amcharts3-angular.directive';
import { AmChartsService } from './amcharts3-angular.service';
import * as i0 from "@angular/core";
export class AmChartsModule {
}
AmChartsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.11", ngImport: i0, type: AmChartsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AmChartsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.11", ngImport: i0, type: AmChartsModule, declarations: [AmChartsDirective], exports: [AmChartsDirective] });
AmChartsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.11", ngImport: i0, type: AmChartsModule, providers: [
        AmChartsService
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.11", ngImport: i0, type: AmChartsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        AmChartsDirective
                    ],
                    exports: [
                        AmChartsDirective
                    ],
                    providers: [
                        AmChartsService
                    ]
                }]
        }] });
